18-pipes.Rmd

```diff
@@ -2,13 +2,136 @@
+- If you use ***ANY*** tidyverse libraries, `{magrittr}` is already included (how to validate this statement)
+- For learning humans... read the `%>%` operator as semantically: *and then*
 
-## Slide 2
+```{r, Loading magrittr, eval=FALSE}
+library(magrittr)
+```
 
-Slide contents.
+
+## Lets see it in Code!!!
+
+The following example will highlight a code snippet **without** piping (Object Assignment and Nested) and **with** piping.
+
+```{r, Example without Pipe, eval = FALSE}
+a <- dplyr::filter(mtcars, carb > 1)
+b <- dplyr::group_by(a, cyl)
+c <- dplyr::summarise(b, Avg_mpg = mean(mpg))
+d <- dplyr::arrange(c, desc(Avg_mpg))
+print(d)
+```
+Above, objects `a-d` are created as "placeholders". Objects that are nothing more then memory space to hold the output of a function. The difference is, each of these can be called upon in future script.
+
+> One of my favorite talks: [Jenny Bryant at Plotcon, 2016](https://youtu.be/4MfUCX_KpdE). Specific to the topic, start at min 15:00. The discussion is related to Dataframes and the key takeaway is wrangle your data once.
+
+But couldn't I just nest my function?
+```{r, Nested Call, eval = FALSE}
+dplyr::arrange(
+   dplyr::summarize(
+       dplyr::group_by(
+           dplyr::filter(mtcars, carb > 1),
+           cyl
+       ),
+       Avg_mpg = mean(mpg)
+   ),
+   desc(Avg_mpg)
+ )
+```
+Nesting a long function is acceptable and in some languages preferred. However, given the structure of RStudio and its plethora of packages, there is no need.
+
+Using the Pipe
+```{r, Example with Pipe, eval=FALSE}
+mtcars %>%
+        dplyr::filter(carb > 1) %>%
+        dplyr::group_by(cyl) %>%
+        dplyr::summarise(Avg_mpg = mean(mpg)) %>%
+        dplyr::arrange(desc(Avg_mpg))
+```
+The third example not only reduces complexity... you are building a pipeline to process the `mtcars` dataset. There is no difference in computation from the first example... only, now everything is in memory.
+
+>NOTE: There is a section in here about `pryr` that I wasn't sure the intent other then a discussion on memory use. R is intellegent and will share colums across dataframes. The material was excluded from the discussion.
+
+## Overwritting Data... easy, right?
+
+One method of wrangling data is to just overwrite and reassign a calculation to its original object. But, there is a problem...
+
+1. Overwriting an object prevents clean debugging. You'll have to rerun your code over and over to see what is changing.
+2. The repetition of the object being transformed implies... "what does my data look like now?"
+
+If you want to make yourself crazy... continually read in a new copy of your data just to walk through the process over agian. This is *not* conducive to efficiency.
+
+## The pipe makes lexical transformations
+
+- `magrittr` reassembles the code in the pipe to a form that works by overwriting an intermediate object, see [link]()
```

Comment:
Where is this link supposed to point?

Reply:
Shoot! I was off on a tangent to find Lexical Transformation/Scoping....I hadn't finished my thought (with the link!). Referencing Chapter 6 of Advanced R. https://adv-r.hadley.nz/functions.html#lexical-scoping. I'll make the change and resubmit.